### PR TITLE
More tweaks and fixes

### PR DESCRIPTION
**Pull request by SiverDX**
WIP

# Config
- Avoid unnecessary config updates
- Don't defer config saving to a new IO thread (seems to lead to more issues)
- Avoid crashes regarding numeric / enum configs
- Check if config entry is valid while editing it and don't allow invalid config entries
- Resource config entries now always show their icons (if the relevant mod is loaded), not just when they're shown as suggestions
- Show resource config icons and suggestions for block, entity type, mob effect and biome configs (was already there but not really used)
- Avoid right click crash (due to adding / removing ui elements instead of setting their visibility)
- Avoid suggestions window overlaps when a new resource entry is focused
- Avoid suggestion entries overlapping the suggestion item tool tip
- Avoid suggestion entries being overlayed by resource entries
- Suggestions can be scrolled through
- Fix some configs not being correctly synced / saved
- Disabled progress slider for numeric values (now always uses a text field)
  - Increments do not work correctly
  - Updates / saves too often causing problems and crashes (might need to add a timer or just save once when the config screen is closed)
- Configs of the type string (non-list) now also show up in the config menu 
- (Probably some other things I forgot)

# Other
- Only load regions_unexplored charged coal recipe when the mod is present
- Avoid hunters_goal tag being broken when zombiemods mod is not present
- Also avoid river biomes for prince / princess spawn (`#forge:is_water`)
- Added config to use vanilla experience bar / food level hud
- Added config to not hide tooltips for unsafe food (food with negative effects)
- Rebuilding of the food map now listens to both Loading and Reloading config events
  - In case the event does not get triggered on config changes through the dragon survival config editor it is also rebuilt directly
- Saturation for food config now allows decimal values
- Added proxy class to call client related code in the network packets to make it more safe
- Fixed dragon claw menu (sometimes?) not being rendered when left open while rejoining
- Fixed potential issue where items could be moved into the crafting result slot while quick moving
- Quick move into the dragon claw tool slots is now possible (while the claw tab is open)
- more comments / cleanup

### CODE REVIEW

**Comment by SiverDX** [comment]
Should be somewhat stable so if anyone wants to test it

File: src/main/java/by/dragonsurvivalteam/dragonsurvival/client/gui/SkinsScreen.java [hunk 1]
```diff
@@ -50,7 +50,7 @@ public class SkinsScreen extends Screen{
 	private static final ResourceLocation WIKI = new ResourceLocation(DragonSurvivalMod.MODID, "textures/gui/wiki_button.png");
 	private static final ResourceLocation HELP = new ResourceLocation(DragonSurvivalMod.MODID, "textures/gui/skin_help.png");
 
-	private static final String DISCORD_URL = "http://discord.gg/8SsB8ar";
+	private static final String DISCORD_URL = "https://discord.gg/8SsB8ar";
```
Comment (BlackAures1) [inline src/main/java/by/dragonsurvivalteam/dragonsurvival/client/gui/SkinsScreen.java:53]: Is there any impact in terms of defensiveness here? The Discord is supposedly already protected

Comment (SiverDX) [reply src/main/java/by/dragonsurvivalteam/dragonsurvival/client/gui/SkinsScreen.java:53]: saves you a redirect to the https url
image

**Comment by BlackAures1** [comment]
> Updates / saves too often causing problems and crashes (might need to add a timer or just save once when the config screen is closed)

Saving after closing the menu is a good idea

> Fixed dragon claw menu (sometimes?) not being rendered when left open while rejoining

I'm really glad someone finally found a way to fix it!

> Quick move into the dragon claw tool slots is now possible (while the claw tab is open)

It's very useful

**Comment by SiverDX** [comment]
is the player wearing any gravitiy modifying equipment (e.g. apotheosis)?

**Comment by SiverDX** [comment]
not sure really
are other parts also affected? since the tail_motion_up value seems to be used for other things as well

**Comment by SiverDX** [comment]
it looks like it gets clamped to the max. value (-10)

``` java
dragon.tailMotionUp = Mth.clamp(Mth.lerp(0.25, dragon.tailMotionUp, ServerFlightHandler.isFlying(player) ? 0 : (player.getDeltaMovement().y + gravity) * 50), -10, 10);
```

so it's either the gravity value (forge attribute) or the delta movement (handled by minecraft)
You could try to add an additional `player.isOnGround()` check and only apply `gravity` in that case
(Did you check if their `y` speed is 0 or assume that since they're on the ground?)


Also, does this always occur? Or just in some cases?
Maybe you weren't in range for the `SyncFlightSpeed` packet

**Comment by BlackAures1** [comment]
To test this bug in the client config, the tail physics must be enabled. We have long tried to add this parameter for more cool animations (like [here](https://youtu.be/lPG6ZnOVqwo)), which react to the direction of movement of the player and make the animation alive. But when trying to use query in molang animations geckolib, it causes visual bugs in multiplayer. Unfortunately, Ais and I don't know how to solve this problem. 

And yes, this graphical bug appears whenever another player jumps/flies and somehow triggers gravity from query. Here's an example of how it's triggered by flight  (look at tail):

https://github.com/DragonSurvivalTeam/DragonSurvival/assets/43911230/0c3e032d-6f24-43f5-a719-f6640cb8ac10

**Comment by SiverDX** [comment]
could be sync issue 

you can try making a test build where you only use the delta movement Y value for the calculation when the dragon is not on the ground and see if it still happens

**Comment by SiverDX** [comment]
are you using `query.tail_motion_up`?

**Comment by SiverDX** [comment]
changed it a bit
tested it in singleplayer and in a multiplayer world and did not have these problems

**Comment by BlackAures1** [comment]
There's no problem in the water?

**Pull request merged**